File: 238.go
package main
import (
    "fmt"
)

func productExceptSelf(nums []int) []int {
    ans := make([]int,len(nums))
    prefix, postfix := 1,1

    for i,_ := range(nums) {
        ans[i]=prefix
        prefix *= nums[i]
    }
    fmt.Println(ans)

    for i:=len(nums)-1; i >=0; i-- {
        ans[i]=postfix * ans[i]
        postfix *= nums[i]
    }
    fmt.Println(ans)

    return ans
}

func main() {
    nums := []int{1,2,3,4}
    ans := productExceptSelf(nums)
    fmt.Println(ans)
}

File: 238.py
# Time O(n)
# Space O(n)
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        postfix = [0]*n
        postfix[-1]=1
        pre = 1
        for i in range(n-2, -1, -1):
            postfix[i] = postfix[i+1]*nums[i+1]
        for i in range(1,n):
            pre = nums[i-1]*pre
            postfix[i]*=pre

        return postfix

